core/trino-main/src/main/java/io/trino/execution/scheduler/BinPackingNodeAllocatorService.java

```diff
@@ -0,0 +1,533 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.scheduler;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Multimap;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import io.airlift.log.Logger;
+import io.trino.Session;
+import io.trino.connector.CatalogName;
+import io.trino.metadata.InternalNode;
+import io.trino.spi.QueryId;
+import io.trino.spi.TrinoException;
+import org.assertj.core.util.VisibleForTesting;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static io.trino.spi.StandardErrorCode.NO_NODES_AVAILABLE;
+import static java.lang.Thread.currentThread;
+import static java.util.Comparator.comparing;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Predicate.not;
+
+public class BinPackingNodeAllocatorService
+        implements NodeAllocatorService
+{
+    private static final Logger log = Logger.get(BinPackingNodeAllocatorService.class);
+
+    @VisibleForTesting
+    static final int PROCESS_PENDING_ACQUIRES_DELAY_SECONDS = 5;
+    private static final int MAX_TASKS_PER_NODE = 1; // TODO be smart here (memory based)
```

Comment:
This is now global, what effectively means only a single task will be running on a node even if multiple queries are running in parallel. Ideally yes, it would be great if the `NodeRequirement` contained a memory requirement so we can bin-pack tasks based on memory (to do that the node information should also probably have the total memory available on a node).

Reply:
1 is a mistake, I think it is an artifact from some manual testing. Ultimately bin-packing approach would be great. For now, I will try to restructure that in a way that:
* the number of slots per node is configurable.
* memory requirement for a task will be passed via `NodeRequirements` object
* if `NodeRequirements.memoryRequrement` is not set then I will assume the task requires a single slot on the node 
* if `NodeRequirements.memoryRequrement` is set I will assert it is set for a huge value and will require a full node for the task.

makes sense @arhimondr ?